src/Sarif.Viewer.VisualStudio/Tags/SarifLocationTagger.cs

```diff
@@ -0,0 +1,573 @@
+        }
+
+        /// <inheritdoc/>
+        public void RemoveTag(ISarifLocationTag tag)
+        {
+            if (!(tag is SarifLocationTag sarifTag))
+            {
+                return;
+            }
+
+            using (this.Update())
+            {
+                using (TagListLock.EnterWriteLock())
+                {
+                    if (this.sarifTags.Remove(sarifTag))
+                    {
+                        this.UpdateBatchSpan(sarifTag.DocumentPersistentSpan.Span);
+
+                        // We do not need TryGetValue here because if it exists in the SARIF tags list
+                        // it must exist in the run Id to SARIF tag map. If it doesn't, it means the lists
+                        // are out of sync which should never happen and is bad.
+                        RunIdToSarifTags[sarifTag.RunId].Remove(sarifTag);
+
+                        sarifTag.PropertyChanged -= this.SarifTagPropertyChanged;
+                        sarifTag.Dispose();
+                    }
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public void RemoveTagsForRun(int runId)
+        {
+            List<SarifLocationTag> tagsToRemove;
+
+            using (TagListLock.EnterReadLock())
+            {
+                if (!RunIdToSarifTags.TryGetValue(runId, out List<SarifLocationTag> sarifTagsForRun) ||
+                    sarifTagsForRun.Count == 0)
+                {
+                    return;
+                }
+
+                // Copy so we can update (which can make outgoing calls) outside of lock
+                tagsToRemove = new List<SarifLocationTag>(sarifTagsForRun);
+            }
+
+            using (this.Update())
+            {
+                foreach (SarifLocationTag tagToRemove in tagsToRemove)
+                {
+                    this.RemoveTag(tagToRemove);
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public IEnumerable<ITagSpan<TextMarkerTag>> GetTags(NormalizedSnapshotSpanCollection spans)
+        {
+            if (spans.Count == 0)
+            {
+                yield break;
+            }
+
+            SarifLocationTag[] possibleTags = null;
+            using (TagListLock.EnterReadLock())
+            {
+                if (this.sarifTags == null)
+                {
+                    yield break;
+                }
+
+                possibleTags = new SarifLocationTag[this.sarifTags.Count];
+                this.sarifTags.CopyTo(possibleTags, 0);
+            }
+
+            if (possibleTags == null)
+            {
+                yield break;
+            }
+
+            foreach (var span in spans)
+            {
+                foreach (var possibleTag in possibleTags.Where((possibleTag) => possibleTag.DocumentPersistentSpan.Span != null))
+                {
+                    SnapshotSpan possibleTagSnapshotSpan = possibleTag.DocumentPersistentSpan.Span.GetSpan(span.Snapshot);
```

Comment:
>GetSpan [](start = 99, length = 7)

Add comment: `GetSpan` maps original span location to new location based on any edits.